server/src/main/java/org/apache/kafka/server/ClientQuotaManager.java

```diff
@@ -0,0 +1,908 @@
+    }
+
+    private MetricConfig getQuotaMetricConfig(double quotaLimit) {
+        return new MetricConfig()
+                .timeWindow(config.quotaWindowSizeSeconds, TimeUnit.SECONDS)
+                .samples(config.numQuotaSamples)
+                .quota(new Quota(quotaLimit, true));
+    }
+
+    protected Sensor getOrCreateSensor(String sensorName, long expirationTimeSeconds, Consumer<Sensor> registerMetrics) {
+        return sensorAccessor.getOrCreate(
+                sensorName,
+                expirationTimeSeconds,
+                registerMetrics);
+    }
+
+    /**
+     * Overrides quotas for <user>, <client-id> or <user, client-id> or the dynamic defaults
+     * for any of these levels.
+     *
+     * @param userEntity   user to override if quota applies to <user> or <user, client-id>
+     * @param clientEntity sanitized client entity to override if quota applies to <client-id> or <user, client-id>
+     * @param quota        custom quota to apply or None if quota override is being removed
+     */
+    public void updateQuota(
+            Optional<ClientQuotaEntity.ConfigEntity> userEntity,
+            Optional<ClientQuotaEntity.ConfigEntity> clientEntity,
+            Optional<Quota> quota
+    ) {
+        /*
+         * Acquire the write lock to apply changes in the quota objects.
+         * This method changes the quota in the overriddenQuota map and applies the update on the actual KafkaMetric object (if it exists).
+         * If the KafkaMetric hasn't been created, the most recent value will be used from the overriddenQuota map.
+         * The write lock prevents quota update and creation at the same time. It also guards against concurrent quota change
+         * notifications
+         */
+        lock.writeLock().lock();
+        try {
+            var quotaEntity = new KafkaQuotaEntity(userEntity.orElse(null), clientEntity.orElse(null));
+
+            // Apply quota changes with proper quota type tracking
+            if (quota.isPresent()) {
+                updateQuotaTypes(quotaEntity, true);
+                quotaCallback.updateQuota(clientQuotaType, quotaEntity, quota.get().bound());
+            } else {
+                updateQuotaTypes(quotaEntity, false);
+                quotaCallback.removeQuota(clientQuotaType, quotaEntity);
+            }
+
+            // Determine which entities need metric config updates
+            Optional<KafkaQuotaEntity> updatedEntity;
+            if (userEntity.filter(entity -> entity == DEFAULT_USER_ENTITY).isPresent() ||
+                    clientEntity.filter(entity -> entity == DEFAULT_USER_CLIENT_ID).isPresent()) {
+                // More than one entity may need updating, so updateQuotaMetricConfigs will go through all metrics
+                updatedEntity = Optional.empty();
+            } else {
+                updatedEntity = Optional.of(quotaEntity);
+            }
+
+            updateQuotaMetricConfigs(updatedEntity);
```

Comment:
Since `updateQuotaMetricConfigs` iterates through all metrics, I'm unsure why it must be executed under a write lock. Perhaps we could create a follow-up ticket for this.

Reply:
open https://issues.apache.org/jira/browse/KAFKA-19475